source/adapters/cuda/command_buffer.cpp

```diff
@@ -401,6 +401,8 @@ urCommandBufferReleaseExp(ur_exp_command_buffer_handle_t hCommandBuffer) {
 
 UR_APIEXPORT ur_result_t UR_APICALL
 urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t hCommandBuffer) {
+  UR_ASSERT(!hCommandBuffer->IsFinalized,
+            UR_RESULT_ERROR_INVALID_COMMAND_BUFFER_EXP);
```

Comment:
Slight preference for `UR_RESULT_ERROR_INVALID_OPERATION` to be the error code to match this OpenCL error code from [clFinalizeCommandBufferKHR](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clFinalizeCommandBufferKHR)

> [CL_INVALID_OPERATION](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_INVALID_OPERATION) if command_buffer is not in the [Recording](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#recording) state.

We should also specify that this error will be returned here https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/exp-command-buffer.yml#L331

Reply:
Nice one, expect you'll need to run `ninja generate` now you've touch yml

Reply:
Good catch. Have updated autogenerated stuff.